nfp/preprocessing/xtb_preprocessor.py

```diff
@@ -0,0 +1,131 @@
+from typing import Dict, Optional, Callable, Hashable
+
+import networkx as nx
+import numpy as np
+import rdkit.Chem
+import json
+
+try:
+    import tensorflow as tf
+except ImportError:
+    tf = None
+
+from nfp.preprocessing import features
+from nfp.preprocessing.preprocessor import Preprocessor
+from nfp.preprocessing.tokenizer import Tokenizer
+
+class xTBPreprocessor(Preprocessor):
+    def __init__(
+            self,
+            atom_features: Optional[Callable[[rdkit.Chem.Atom], Hashable]] = None,
+            bond_features: Optional[Callable[[rdkit.Chem.Bond], Hashable]] = None,
+            **kwargs,
+    ) -> None:
+        super(xTBPreprocessor, self).__init__(**kwargs)
+
+        #first set of RDKit features for keeping count of type of atom/bond features.
+        if atom_features is None:
+            self.atom_features_xtb = features.atom_features_xtb
+        if bond_features is None:
+            self.bond_features_xtb = features.bond_features_xtb
+
+        self.atom_tokenizer = Tokenizer()
+        self.bond_tokenizer = Tokenizer()
+
+
+    def create_nx_graph(self, mol: rdkit.Chem.Mol, **kwargs) -> nx.DiGraph:
+        #add hydrogens as wbo contains Hydrogens
+        g = nx.Graph(mol=mol)
+        g.add_nodes_from(((atom.GetIdx(), {
+            'atom': atom
+        }) for atom in mol.GetAtoms()))
+
+        #add edges based on wilberg bond orders.
+        wbo = np.array(self.json_data['Wiberg matrix'])
+        g.add_edges_from(((i, j, {
+            'wbo': wbo[i][j],'mol': mol
+        }) for i in range(len(wbo)) for j in range(len(wbo)) if wbo[i][j] > self.cutoff ))
+
+        return nx.DiGraph(g)
+
+    def get_edge_features(self, edge_data: list,
+                          max_num_edges) -> Dict[str, np.ndarray]:
+        bond_feature_matrix = np.zeros(max_num_edges, dtype=self.output_dtype)
+        for n, (start_atom, end_atom, bond_dict) in enumerate(edge_data):
+
+            # flipped = start_atom == bond_dict['bond'].GetEndAtomIdx()
+            bond_feature_matrix[n] = self.bond_tokenizer(
+                self.bond_features_xtb(start_atom, end_atom, bond_dict['wbo'],bond_dict['mol']))
+        return {'n_bond': max_num_edges,'bond': bond_feature_matrix}
```

Comment:
part of the reason for breaking this up was that you could have additional edge features, i.e., besides just bond class. so you could return 'bond', 'bond_lengths' as different properties, and handle the float features separately in your neural network.

Reply:
Done